Sources/PT.PM.Common/MessagePack/RootUstFormatter.cs

```diff
@@ -0,0 +1,146 @@
+using System;
+using System.Collections.Generic;
+using MessagePack;
+using MessagePack.Formatters;
+using PT.PM.Common.Exceptions;
+using PT.PM.Common.Files;
+using PT.PM.Common.Nodes;
+using PT.PM.Common.Nodes.Tokens.Literals;
+
+namespace PT.PM.Common.MessagePack
+{
+    public class RootUstFormatter : IMessagePackFormatter<RootUst>
+    {
+        public RootUst CurrentRoot { get; internal set; }
+
+        public BinaryFile SerializedFile { get; private set; }
+
+        public HashSet<IFile> SourceFiles { get; private set; }
+
+        public static RootUstFormatter CreateWriter()
+        {
+            return new RootUstFormatter();
+        }
+
+        public static RootUstFormatter CreateReader(BinaryFile serializedFile, HashSet<IFile> sourceFiles)
+        {
+            return new RootUstFormatter
+            {
+                SerializedFile = serializedFile ?? throw new ArgumentNullException(nameof(serializedFile)),
+                SourceFiles = sourceFiles ?? throw new ArgumentNullException(nameof(sourceFiles))
+            };
+        }
+
+        private RootUstFormatter()
+        {
+        }
+
+        public int Serialize(ref byte[] bytes, int offset, RootUst value, IFormatterResolver formatterResolver)
+        {
+            int newOffset = offset;
+            CurrentRoot = value;
+
+            var languageFormatter = formatterResolver.GetFormatter<Language>();
+            newOffset += languageFormatter.Serialize(ref bytes, newOffset, value.Language, formatterResolver);
+
+            var localSourceFiles = new List<TextFile> {CurrentRoot.SourceFile};
+            value.ApplyActionToDescendantsAndSelf(ust =>
+            {
+                foreach (TextSpan textSpan in ust.TextSpans)
+                {
+                    if (textSpan.File != null && !localSourceFiles.Contains(textSpan.File))
+                    {
+                        localSourceFiles.Add(textSpan.File);
+                    }
+                }
+            });
+
+            var textSpanFormatter = formatterResolver.GetFormatter<TextSpan>();
+            if (textSpanFormatter is TextSpanFormatter formatter)
```

Comment:
Are there cases when in this place the formatter is not a `TextSpanFormatter`?

Reply:
No. But I like such a style, because it's safer in general.